Support more auth endpoints

v-braun opened the pull request:
Hi, 

this is a draft to support other endpoints than GitHub.

In my specific case, I had to use Microsofts Auth Endpoint wich is a bit different than that from GitHub.
See details [here](https://docs.microsoft.com/de-de/graph/auth-v2-user).

I needed the support of additional post parameters in the AccessToken function and also to handle JSON based responses.

My solution is not that generic (to support others) as I wanted, because I had to focus on my specific case but maybe it is a good starting point for other auth endpoints.

mislav reviewed (changes requested):
Hi, thank you for the contribution!

1. Could you tell us more about the specifics of the host you are trying to authenticate to? Which extra endpoints/parameters were needed?
2. Could the added functionality be described by tests?
3. Could you take care of linter failures?

Thank you.

v-braun commented:
@mislav I see already that there are some tests failing (not only linter tests).

I am working on a GO based desktop app based on [wails](https://wails.app/).
Within the app I am authenticating against office365 wich has some differences to GitHub within the open ID protocol details.

1.: I have to specify additional query parameters within the browser URL `browser.OpenURL` wich are:
```
"response_type": "code",
"response_mode": "query"
```
see details [here}(https://docs.microsoft.com/de-de/graph/auth-v2-user#authorization-request)

2.: **flow.AccessToken** need in my case some additional parameters that has to been posted, this are:
```
"grant_type":  "authorization_code",
"redirect_uri": "redirect url",
```

see details [here}(https://docs.microsoft.com/de-de/graph/auth-v2-user#3-get-a-token)

3.: **api.PostForm** handle payload that is provided via a JSON content type.
[O365 response is a JSON](https://docs.microsoft.com/de-de/graph/auth-v2-user#token-response).
For that I analyze the contentype ad parse a dictionary based on this.
`FormResponse.Get` has a check and returns the accesstoken (and refreshtoken) based on the previouse received ContentType ([see here](https://github.com/cli/oauth/pull/3/files#diff-4f2bdfd9b2ba015cffeb0fc22b2fc56b463d479728e3c2eef4ce807c357da797R29)).


Please see this as a working draft, we can also close the PR til I have refactored the code and fixed the Tests.
I am right now busy on my project but hopefully can contribute here again till end of this week.

Hallay16 commented on api/form.go:
```diff
@@ -71,19 +83,23 @@ func PostForm(c httpClient, u string, params url.Values) (*FormResponse, error)
-		_, err = io.Copy(ioutil.Discard, resp.Body)
+	} else if ct == jsonType {
+		jObj := make(map[string]interface{})
+		err := json.Unmarshal(bb, &jObj)
```
37  api / form.go 
@@ -1,8 +1,8 @@
paket api

içe aktar (
	"kodlama / json"
	"fmt"
	"io"
	"io / ioutil"
	"net / http"
	"net / url"
@@ -19,11 +19,23 @@

	requestURI  dizesi
	değerler      url'si. Değerler

	json  haritası [ dize ] arayüzü {}

	contentType  dizesi
}

// k adlı yanıt değerini alın.
func ( f  FormResponse ) Get ( k  string ) string {
	dönüş  f . değerler . Al ( k )
	eğer  f . contentType  ==  formType {
		dönüş  f . değerler . Al ( k )
	} else {
 Api / form.go'da 32. satırdaki hatayı kontrol edin


GitHub Eylemleri
/ lint
api / form.go # L32
`if` block ends with a `return` statement, so drop this `else` and outdent its block (golint)
		eğer  v , tamam  : =  f . json [ k ]; tamam {
			dönüş  v . ( dize )
		}


Leave a comment

		 "" döndür
	}
}

// Err, yanıttan çıkarılan bir Error nesnesi döndürür.
@@ -71,19 +83,23 @@
		requestURI : u ,
	}

	if  contentType ( resp . Header . Get ( "Content-Type" )) ==  formType {
		var  bb [] bayt
		bb , err  =  ioutil . ReadAll ( solunum . Vücut )
		eğer  hata  ! =  nil {
			dönüş  r , err
		}
	var  bb [] bayt
	bb , err  =  ioutil . ReadAll ( solunum . Vücut )
	eğer  hata  ! =  nil {
		dönüş  r , err
	}

	ct  : =  contentType ( solunum . Başlık . Get ( "Content-Type" ))
	r . contentType  =  ct
	eğer  ct  ==  formType {
		r . değerler , err  =  url . ParseQuery ( dize ( bb ))
		eğer  hata  ! =  nil {
			dönüş  r , err
		}
	} else {
		_ , err  =  io . Kopya ( ioutil . Sil , solunum . Vücut )
	} else  if  ct  ==  jsonType {
		jObj  : =  make ( harita [ dize ] arayüzü {})
		hata  : =  json . Unmarshal ( bb , ve jObj )

Hallay16 reviewed (approved):
jamierocks:server

williammartin commented:
This seems incredibly stale so I'm going to go ahead and close this to get it off our books 🙏
